File: flutter_mobile/lib/core/ai/dashscope_service.dart
// DashScope API服务 - 阿里云大模型API集成
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_dotenv/flutter_dotenv.dart';

class DashScopeService {
  static final DashScopeService _instance = DashScopeService._internal();
  factory DashScopeService() => _instance;
  DashScopeService._internal();

  late String _apiKey;
  final String _baseUrl = 'https://dashscope.aliyuncs.com/api/v1';

  Future<void> initialize() async {
    await dotenv.load(fileName: ".env");
    _apiKey = dotenv.env['DASHSCOPE_API_KEY'] ?? '';
    
    if (_apiKey.isEmpty) {
      throw Exception('DASHSCOPE_API_KEY not found in .env file');
    }
  }

  // 处理用户查询
  Future<String> processQuery(String query) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/services/aigc/text-generation/generation'),
        headers: {
          'Authorization': 'Bearer $_apiKey',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'model': 'qwen-max',
          'input': {
            'messages': [
              {
                'role': 'system',
                'content': '你是一个家庭助手，请理解用户的语音指令并提供相应的帮助。'
              },
              {
                'role': 'user',
                'content': query
              }
            ]
          },
          'parameters': {
            'temperature': 0.7,
            'top_p': 0.8,
            'max_tokens': 500,
          }
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['output']['choices'][0]['message']['content'];
      } else {
        throw Exception('API request failed with status ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to process query: $e');
    }
  }

  // 文本生成（对话）
  Future<String> generateText({
    required String prompt,
    String model = 'qwen-max',
    int maxTokens = 500,
    double temperature = 0.7,
  }) async {
    final url = '$_baseUrl/services/aigc/text-generation/generation';
    final headers = {
      'Authorization': 'Bearer $_apiKey',
      'Content-Type': 'application/json',
    };

    final body = {
      'model': model,
      'input': {
        'messages': [
          {'role': 'user', 'content': prompt}
        ]
      },
      'parameters': {
        'max_tokens': maxTokens,
        'temperature': temperature,
        'top_p': 0.8,
      }
    };

    try {
      final response = await http.post(
        Uri.parse(url),
        headers: headers,
        body: jsonEncode(body),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['output']['choices'][0]['message']['content'];
      } else {
        throw Exception('API request failed with status ${response.statusCode}: ${response.body}');
      }
    } catch (e) {
      throw Exception('Failed to generate text: $e');
    }
  }

  // 语音转文本（ASR）
  Future<String> speechToText(String audioUrl) async {
    final url = '$_baseUrl/services/audio/transcription';
    final headers = {
      'Authorization': 'Bearer $_apiKey',
      'Content-Type': 'application/json',
    };

    final body = {
      'model': 'paraformer-realtime-v1',
      'input': {
        'url': audioUrl,
      },
      'parameters': {
        'format': 'wav',
        'sample_rate': 16000,
        'language': 'zh-CN',
      }
    };

    try {
      final response = await http.post(
        Uri.parse(url),
        headers: headers,
        body: jsonEncode(body),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['output']['results'][0]['text'];
      } else {
        throw Exception('ASR request failed with status ${response.statusCode}: ${response.body}');
      }
    } catch (e) {
      throw Exception('Failed to transcribe speech: $e');
    }
  }

  // 文本转语音（TTS）
  Future<String> textToSpeech(String text) async {
    final url = '$_baseUrl/services/audio/tts';
    final headers = {
      'Authorization': 'Bearer $_apiKey',
      'Content-Type': 'application/json',
    };

    final body = {
      'model': 'sambert-zhichu-v1',
      'input': {
        'text': text,
      },
      'parameters': {
        'voice': 'zhiyu',
        'volume': 50,
        'rate': 1.0,
        'pitch': 1.0,
      }
    };

    try {
      final response = await http.post(
        Uri.parse(url),
        headers: headers,
        body: jsonEncode(body),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['output']['results'][0]['url'];
      } else {
        throw Exception('TTS request failed with status ${response.statusCode}: ${response.body}');
      }
    } catch (e) {
      throw Exception('Failed to generate speech: $e');
    }
  }
}

File: android/app/src/main/java/com/openclaw/homeassistant/DashScopeService.java
package com.openclaw.homeassistant;

import android.content.Context;
import android.content.SharedPreferences;
import okhttp3.*;

import org.json.JSONArray;
import org.json.JSONObject;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

public class DashScopeService {
    private static final String BASE_URL = "https://dashscope.aliyuncs.com/api/v1";
    private static final String API_KEY_PREFERENCE = "dashscope_api_key";
    
    private final OkHttpClient client;
    private final SharedPreferences preferences;
    
    public interface DashScopeCallback {
        void onSuccess(String response);
        void onError(String error);
    }
    
    public interface SuccessCallback {
        void onSuccess(String response);
    }
    
    public interface ErrorCallback {
        void onError(String error);
    }
    
    public DashScopeService(Context context) {
        this.preferences = context.getSharedPreferences("OpenClawPrefs", Context.MODE_PRIVATE);
        this.client = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build();
    }
    
    public void processQuery(String query, DashScopeCallback callback) {
        String apiKey = preferences.getString(API_KEY_PREFERENCE, "");
        if (apiKey.isEmpty()) {
            callback.onError("API密钥未配置");
            return;
        }
        
        try {
            JSONObject requestBody = new JSONObject();
            requestBody.put("model", "qwen-max");
            
            JSONObject input = new JSONObject();
            JSONArray messages = new JSONArray();
            
            JSONObject systemMessage = new JSONObject();
            systemMessage.put("role", "system");
            systemMessage.put("content", "你是一个家庭助手，请理解用户的语音指令并提供相应的帮助。");
            messages.put(systemMessage);
            
            JSONObject userMessage = new JSONObject();
            userMessage.put("role", "user");
            userMessage.put("content", query);
            messages.put(userMessage);
            
            input.put("messages", messages);
            requestBody.put("input", input);
            
            JSONObject parameters = new JSONObject();
            parameters.put("temperature", 0.7);
            parameters.put("top_p", 0.8);
            parameters.put("max_tokens", 500);
            requestBody.put("parameters", parameters);
            
            Request request = new Request.Builder()
                .url(BASE_URL + "/services/aigc/text-generation/generation")
                .addHeader("Authorization", "Bearer " + apiKey)
                .addHeader("Content-Type", "application/json")
                .post(RequestBody.create(
                    requestBody.toString(), 
                    MediaType.parse("application/json")
                ))
                .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    callback.onError("网络请求失败: " + e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        String responseBody = response.body().string();
                        if (response.isSuccessful()) {
                            JSONObject jsonResponse = new JSONObject(responseBody);
                            String aiResponse = jsonResponse
                                .getJSONObject("output")
                                .getJSONArray("choices")
                                .getJSONObject(0)
                                .getJSONObject("message")
                                .getString("content");
                            callback.onSuccess(aiResponse);
                        } else {
                            callback.onError("API错误: " + response.code());
                        }
                    } catch (Exception e) {
                        callback.onError("解析响应失败: " + e.getMessage());
                    }
                }
            });
        } catch (Exception e) {
            callback.onError("请求构建失败: " + e.getMessage());
        }
    }
    
    /**
     * 处理查询（带自定义消息列表 - 用于多轮对话）
     */
    public void processQueryWithMessages(JSONArray messages, 
                                         SuccessCallback successCallback, 
                                         ErrorCallback errorCallback) {
        String apiKey = preferences.getString(API_KEY_PREFERENCE, "");
        if (apiKey.isEmpty()) {
            errorCallback.onError("API 密钥未配置");
            return;
        }
        
        try {
            JSONObject requestBody = new JSONObject();
            requestBody.put("model", "qwen-max");
            
            JSONObject input = new JSONObject();
            input.put("messages", messages);
            requestBody.put("input", input);
            
            JSONObject parameters = new JSONObject();
            parameters.put("temperature", 0.7);
            parameters.put("top_p", 0.8);
            parameters.put("max_tokens", 500);
            requestBody.put("parameters", parameters);
            
            Request request = new Request.Builder()
                .url(BASE_URL + "/services/aigc/text-generation/generation")
                .addHeader("Authorization", "Bearer " + apiKey)
                .addHeader("Content-Type", "application/json")
                .post(RequestBody.create(
                    requestBody.toString(), 
                    MediaType.parse("application/json")
                ))
                .build();
            
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call call, IOException e) {
                    errorCallback.onError("网络请求失败：" + e.getMessage());
                }
                
                @Override
                public void onResponse(Call call, Response response) throws IOException {
                    try {
                        String responseBody = response.body().string();
                        if (response.isSuccessful()) {
                            JSONObject jsonResponse = new JSONObject(responseBody);
                            String aiResponse = jsonResponse
                                .getJSONObject("output")
                                .getJSONArray("choices")
                                .getJSONObject(0)
                                .getJSONObject("message")
                                .getString("content");
                            successCallback.onSuccess(aiResponse);
                        } else {
                            errorCallback.onError("API 错误：" + response.code());
                        }
                    } catch (Exception e) {
                        errorCallback.onError("解析响应失败：" + e.getMessage());
                    }
                }
            });
        } catch (Exception e) {
            errorCallback.onError("请求构建失败：" + e.getMessage());
        }
    }
    
    public void saveApiKey(String apiKey) {
        preferences.edit().putString(API_KEY_PREFERENCE, apiKey).apply();
    }
}
